Expose type parameter in simple query string

Kiriakos1998 opened the pull request:
Support type parameter for simple_query_string that determines how the query matches and scores documents over multiple fields.
Closes(#58375)

Kiriakos1998 committed:
Expose type parameter in simple query string

Kiriakos1998 committed:
Add changelog

cbuescher reviewed (changes requested):
Thanks @Kiriakos1998 for opening this PR.
I left a couple of comments, the newly introduced property in the builder would need some additional treatment for retaining backward compatibility when serializing to older nodes. Let me come back with some details on that later.
I also notices when running some initial checks locally that e.g. `SimpleQueryStringBuilderTests` is frequently failing the "testDefaults", "testToQuery" and occasionally the "textCachability" methods. This most likely relates to the new randomized type and wouldn't be the same with every test run. Could you check those failures and maybe run at least the "server" module rest (./gradlew -p server test) locally to check?

cbuescher commented:
I just took a look at how we currently handle additions to the transport protocol like in this case in the builder. I think you need to

1. Add a new TransportVersion to the similarly named file, e.g. like in this recent PR https://github.com/elastic/elasticsearch/pull/96953/files#diff-3e9f8961608cb91f073796642945c094c750f59b1f57b0ea7b710a33024ad78c
2. Surround the code that writes and reads from StreamInput / StreamOutput in the corresponding class constructor and the "doWriteTo" methods with a conditional check for these versions like in e.g. [this examle](https://github.com/elastic/elasticsearch/pull/96953/files#diff-cc94b6c97b605e29bde7bb31a0d4af9b1a368818e7261c48a34590c93195ee8eR92)

There might be a few more steps involved. Maybe @thecoop has more input on that or can point to a public README or document that explains this in more detail.

thecoop commented:
Yes, that's pretty much it. We need to add it to the user docs about adding new transport versions (#97050)

Essentially, whenever you make a change to how an object is serialized, it needs to be conditional on the negotiated transport version between two nodes.

So, do the following:
1. Add a new sequential constant to `TransportVersion.java` for your change (eg https://github.com/elastic/elasticsearch/pull/96004/files#diff-3e9f8961608cb91f073796642945c094c750f59b1f57b0ea7b710a33024ad78c)
2. Make your serialization changes conditional on the stream version being at least that version (eg https://github.com/elastic/elasticsearch/pull/96004/files#diff-1012fda7247125ee9c8155576421bd29a6a54250cf3bbd846055d4532c8020cbR168)
3. Add a default when talking to nodes that don't understand the most recent change.

Kiriakos1998 committed:
backwards compatibility

Kiriakos1998 commented:
Thanks for the feedback @cbuescher. I did the changes for the compatibility and also addressed some issues in the tests. However, I am not sure how to tackle the issue with the randomizer. At the moment I commented out the specific line and no test fails. Should I delete it or it is needed?

cbuescher commented on server/src/test/java/org/elasticsearch/index/query/SimpleQueryStringBuilderTests.java:
```diff
@@ -108,6 +108,7 @@ protected SimpleQueryStringBuilder doCreateTestQueryBuilder() {
         if (randomBoolean()) {
             result.fuzzyTranspositions(randomBoolean());
         }
+        // result.type(randomFrom(MultiMatchQueryBuilder.Type.values()));
```
I think BEST_FIELDS, MOST_FIELDS and PHRASE work out okay in the "testToQuery" test, the other three might been some work to extend the assertions in "doAssertLuceneQuery" to work. 
One option would be to only draw from the three types that work here, so we at least get _some_ randomization, and maybe leave a TODO to extend the test later. I think given that the test that check the generated Lucene queries for a query builder are brittle anyway, this solution would be okay.
The second option is to go through each of the failing three types (BOOL_PREFIX, PHRASE_PREFIX and CROSS_FIELDS) and extend the checks in "doAssertLuceneQuery" according to what queries are produced in those cases. This is not a small task since there might be a ton of combinations possible (based on other query builder parameters and field types etc.), and the value of having those tests is debateable. Up to you I guess.

Kiriakos1998 replied:
I took the first option since the second was not a small task and you mentioned that the value that would be added was debatable.

cbuescher reviewed (commented):
Thanks for the iteration, I left a small suggestion how I think we can get the test running.

Kiriakos1998 committed:
address testToQuery partial failure

Kiriakos1998 commented on server/src/main/java/org/elasticsearch/TransportVersion.java:
```diff
@@ -147,9 +147,10 @@ private static TransportVersion registerTransportVersion(int id, String uniqueId
     // 8.9.0
     public static final TransportVersion V_8_500_020 = registerTransportVersion(8_500_020, "ECB42C26-B258-42E5-A835-E31AF84A76DE");
     public static final TransportVersion V_8_500_021 = registerTransportVersion(8_500_021, "102e0d84-0c08-402c-a696-935f3a3da873");
+    public static final TransportVersion V_8_500_022 = registerTransportVersion(8_500_022, "01b06435-5d73-42ff-a121-3b36b771375e");
```
Done

thecoop replied:
@cbuescher That's not correct, the comment for 8.10.0 should be added when the 8.10 ES release is frozen in August.

cbuescher replied:
Thanks for the corrections, my bad, good to know for next time. I took the liberty to remove the comment myself.

cbuescher commented on server/src/main/java/org/elasticsearch/TransportVersion.java:
```diff
@@ -150,9 +150,11 @@ private static TransportVersion registerTransportVersion(int id, String uniqueId
     public static final TransportVersion V_8_500_021 = registerTransportVersion(8_500_021, "102e0d84-0c08-402c-a696-935f3a3da873");
     // Introduced for stateless plugin
     public static final TransportVersion V_8_500_022 = registerTransportVersion(8_500_022, "4993c724-7a81-4955-84e7-403484610091");
+    // 8.10.0
```
```suggestion
```

cbuescher committed:
Update server/src/main/java/org/elasticsearch/TransportVersion.java

Removed, sorry my bad

cbuescher commented on server/src/test/java/org/elasticsearch/index/query/SimpleQueryStringBuilderTests.java:
```diff
@@ -179,11 +190,17 @@ public void testDefaultEmptyComplainFlags() {
             qb.defaultOperator()
         );
+        assertEquals(
+            "Setting operator to null should result in returning to default value.",
```
nit: minor typo here
```suggestion
            "Setting type to null should result in returning to default value.",
```

cbuescher committed:
Update server/src/test/java/org/elasticsearch/index/query/SimpleQueryStringBuilderTests.java

I'll commit this small type change myself so I can run tests.

cbuescher reviewed (approved):
Thanks @Kiriakos1998 for the updates, LGTM. I commited two minor changes myself to be able to run final CI tests. Hope you don't mind.

cbuescher commented:
@elasticmachine test this please

Kiriakos1998 commented:
> @elasticmachine test this please

It fails in code style. I run locally spotlessApply. Let me know if I should push now or after the suite finishes.

cbuescher commented:
> in code style. I run locally spotlessApply
Yes please, I tried fixing through Github UI via the "suggest" feature but its tricky. Better to clean if up on your side if you don't mind. Thanks.

Kiriakos1998 committed:
spotlessApply

Kiriakos1998 commented:
> > in code style. I run locally spotlessApply
> 
> Yes please, I tried fixing through Github UI via the "suggest" feature but its tricky. Better to clean if up on your side if you don't mind. Thanks.

Done

cbuescher commented:
@elasticmachine test this please

cbuescher commented:
@Kiriakos1998 merged, thanks a lot for your contribution and working on this PR!
